src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -791,3 +938,312 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+
+	if (entryFound)
+	{
+		if (isModificationQuery)
+		{
+			RecordParticipatingShardId(placement->shardId, participantEntry);
+		}
+
+		return participantEntry->connection;
+	}
+	else
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_DOES_NOT_EXIST),
+						errmsg("no transaction participant matches %s:%d",
+							   placement->nodeName, placement->nodePort),
+						errdetail("Transactions which modify distributed tables may only "
+								  "target nodes affected by the modification command "
+								  "which began the transaction.")));
+	}
+}
+
+
+/*
+ * PurgeConnectionForPlacement provides a way to purge an invalid connection
+ * from all relevant connection hashes using the placement involved in the
+ * query at the time of the error. If a transaction is ongoing, this function
+ * ensures the right node's connection is set to NULL in the participant map
+ * for the transaction in addition to purging the connection cache's entry.
+ */
+static void
+PurgeConnectionForPlacement(ShardPlacement *placement)
+{
+	NodeConnectionKey nodeKey;
+	char *currentUser = CurrentUserName();
+
+	MemSet(&nodeKey, 0, sizeof(NodeConnectionKey));
+	strlcpy(nodeKey.nodeName, placement->nodeName, MAX_NODE_LENGTH + 1);
+	nodeKey.nodePort = placement->nodePort;
+	strlcpy(nodeKey.nodeUser, currentUser, NAMEDATALEN);
+
+	PurgeConnectionByKey(&nodeKey);
+
+	if (xactParticipantHash != NULL)
+	{
+		XactParticipantKey participantKey;
+		XactParticipantEntry *participantEntry = NULL;
+		bool entryFound = false;
+
+		Assert(IsTransactionBlock());
+
+		memcpy(&participantKey, &nodeKey, sizeof(XactParticipantKey));
+		participantEntry = hash_search(xactParticipantHash, &participantKey, HASH_FIND,
+									   &entryFound);
+
+		Assert(entryFound);
+
+		participantEntry->connection = NULL;
+	}
+}
+
+
+/*
+ * RecordParticipatingShardId simply adds a shard identifier to the list of
+ * dirtied shards in a transaction participant entry. Conceptually similar to
+ * list_append_uniq, but with uint64 and memory context handling.
+ */
+static void
+RecordParticipatingShardId(uint64 newShardId, XactParticipantEntry *participant)
+{
+	ListCell *shardCell = NULL;
+	MemoryContext oldContext = NULL;
+	uint64 *newShardIdPtr = NULL;
+
+	foreach(shardCell, participant->shardIds)
+	{
+		uint64 *shardIdPointer = (uint64 *) lfirst(shardCell);
+		uint64 shardId = (*shardIdPointer);
+
+		if (shardId == newShardId)
+		{
+			return;
+		}
+	}
+
+	oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+	newShardIdPtr = AllocateUint64(newShardId);
+	participant->shardIds = lappend(participant->shardIds, newShardIdPtr);
+
+	MemoryContextSwitchTo(oldContext);
+}
+
+
+/*
+ * MarkParticipantShardsUnhealthy marks all of a transaction participant's
+ * shards as inactive. Used after a final COMMIT has failed (otherwise shards
+ * are just marked inactive as each modification command fails).
+ */
+static void
+MarkParticipantShardsUnhealthy(XactParticipantEntry *participant)
+{
```

Comment:
So I think this is a bit of the wrong logic.  We should be marking things as invalid based on the shards, not the transactions. I.e. track which shard have failed on which placements, and mark as invalid iff there more than one placement, and not all placements failed.